gcloud-java-bigquery/src/main/java/com/google/cloud/bigquery/Job.java

```diff
@@ -144,14 +163,51 @@ public boolean isDone() {
   }
 
   /**
+   * Waits until this job completes its execution, either failing or succeeding. If the job does not
+   * exist, this method returns without executing any method of the provided callback. If the job
+   * completed successfully the {@link CompletionCallback#success(Job)} method is called. If the job
+   * completed with errors the {@link CompletionCallback#error(BigQueryError, List)} method is
+   * called.
+   * <pre> {@code
+   * job.whenDone(new CompletionCallback() {
+   *   void success(Job job) {
+   *     // completed successfully
+   *   }
+   *
+   *   void error(BigQueryError error, List<BigQueryError> executionErrors) {
+   *     // handle error
+   *   }
+   * });}</pre>
+   *
+   * @throws BigQueryException upon failure
+   * @throws InterruptedException if the current thread gets interrupted while waiting for the job
+   *     to complete
+   */
+  public void whenDone(CompletionCallback callback) throws InterruptedException {
```

Comment:
This method is blocking. So what should people expect once it completed... They should expect that job is completed (failed or successful). So whatever they do in the callback they can do *after* the blocking call.

If you provide a callback I think it is much more expected that all would be handled in the callback. If that is not the case then I think the following application code would be simpler:

```java
try {
  Job completedJob = job.waitFor();
  if (completedJob == null) {
     // job was removed (if you care, otherwise just check for completedJob != null
  } else if (completedJob.status().error() != null) {
    // job failed,  do something AND also have a reference to the job as you may need it
    // e.g. to log job X failed.
  } else {
    // job sucessful
  }
} catch (ComputeException ex) {
...
}
```

As oppose to:
```java
try {
  job.whenDone(new CompletionCallback() {
    @Override
     public void success(Job job) {
        // if the code after the blocking `whenDone` needs an updated reference
        // I will need to play some trickery to get this reference visible after
     }

     @Override
      public void error(BigQueryError error, List<BigQueryError> executionErrors) {
        // what if I need a reference to the Job?
      }
  });
  
   // now I can also deal with the job and I will only get here once the job was completed
   // however I don't have a reference for the updated job nor do I know if actually the job
   // was removed and non of my callbacks were called. So, if I care about it I have to do:
  job = job.reload();
  if (job = null) {
   ...
  } else {
    ...
  }    
} catch (ComputeException ex) {
...
}
```